Modules/mathmodule.c

```diff
@@ -2031,49 +2031,64 @@ math_fmod_impl(PyObject *module, double x, double y)
         return PyFloat_FromDouble(r);
 }
 
-
-/*[clinic input]
-math.hypot
-
-    x: double
-    y: double
-    /
-
-Return the Euclidean distance, sqrt(x*x + y*y).
-[clinic start generated code]*/
-
+/* AC: cannot convert yet, waiting for *args support */
 static PyObject *
-math_hypot_impl(PyObject *module, double x, double y)
-/*[clinic end generated code: output=b7686e5be468ef87 input=7f8eea70406474aa]*/
+math_hypot(PyObject *self, PyObject *args)
 {
-    double r;
-    /* hypot(x, +/-Inf) returns Inf, even if x is a NaN. */
-    if (Py_IS_INFINITY(x))
-        return PyFloat_FromDouble(fabs(x));
-    if (Py_IS_INFINITY(y))
-        return PyFloat_FromDouble(fabs(y));
-    errno = 0;
-    PyFPE_START_PROTECT("in math_hypot", return 0);
-    r = hypot(x, y);
-    PyFPE_END_PROTECT(r);
-    if (Py_IS_NAN(r)) {
-        if (!Py_IS_NAN(x) && !Py_IS_NAN(y))
-            errno = EDOM;
-        else
-            errno = 0;
+    Py_ssize_t i, n;
+    PyObject *item;
+    double max = 0.0;
+    double csum = 0.0;
+    double x;
+    int found_inf = 0;
+    int found_nan = 0;
+
+    n = PyTuple_GET_SIZE(args);
+    for (i=0 ; i<n ; i++) {
+        item = PyTuple_GET_ITEM(args, i);
+        x = PyFloat_AsDouble(item);
+        if (x == -1.0 && PyErr_Occurred()) {
+            return NULL;
+        }
+        found_inf |= Py_IS_INFINITY(x);
+        found_nan |= Py_IS_NAN(x);
+        x = fabs(x);
+        if (x > max) {
+            max = x;
+        }
     }
-    else if (Py_IS_INFINITY(r)) {
-        if (Py_IS_FINITE(x) && Py_IS_FINITE(y))
-            errno = ERANGE;
-        else
-            errno = 0;
+    if (found_inf) {
+        return PyFloat_FromDouble(max);
     }
-    if (errno && is_error(r))
-        return NULL;
-    else
-        return PyFloat_FromDouble(r);
+    if (found_nan) {
+        return PyFloat_FromDouble(Py_NAN);
+    }
+    if (max == 0.0) {
+        return PyFloat_FromDouble(0.0);
+    }
+    for (i=0 ; i<n ; i++) {
+        item = PyTuple_GET_ITEM(args, i);
+        x = PyFloat_AsDouble(item) / max;
```

Comment:
Are we quite sure a malicious `__float__()` implementation can't mutate other values in the argument tuple?  That is, that no conversion failed the first time through the tuple guarantees none will fail the second time through?

I don't know.  In which case checking for a -1.0 error return here too would be safest.

Reply:
Okay, added the -1.0 test.